File: zy70458/port_inspector/rule_manager.py
import json
from datetime import datetime
from pathlib import Path
from typing import Dict, List, Optional
from .models import PortInspectionRule, RuleVersion, RiskLevel


class RuleManager:
    def __init__(self, data_dir: Path):
        self.data_dir = data_dir
        self.rules_dir = data_dir / "rules"
        self.rules_dir.mkdir(parents=True, exist_ok=True)
        self._init_default_rules()

    def _init_default_rules(self):
        rule_file = self.rules_dir / "rules_v1.0.0.json"
        if not rule_file.exists():
            default_rule = PortInspectionRule(
                version="v1.0.0",
                port_ranges={
                    "0-1023": RiskLevel.CRITICAL,
                    "1024-49151": RiskLevel.MEDIUM,
                    "49152-65535": RiskLevel.LOW,
                },
                reserved_ports=[22, 80, 443, 3306, 5432, 6379, 27017],
                threshold_connections=100,
                description="初始版本规则：系统端口CRITICAL，注册端口MEDIUM，动态端口LOW"
            )
            self._save_rule(default_rule)

        rule_file_v2 = self.rules_dir / "rules_v2.0.0.json"
        if not rule_file_v2.exists():
            v2_rule = PortInspectionRule(
                version="v2.0.0",
                port_ranges={
                    "0-1023": RiskLevel.HIGH,
                    "1024-49151": RiskLevel.LOW,
                    "49152-65535": RiskLevel.SAFE,
                },
                reserved_ports=[22, 80, 443, 3306, 5432, 6379, 27017, 8080, 8443],
                threshold_connections=50,
                description="v2.0版本：降低系统端口等级为HIGH，注册端口为LOW，新增常用端口"
            )
            self._save_rule(v2_rule)

    def _save_rule(self, rule: PortInspectionRule):
        rule_file = self.rules_dir / f"rules_{rule.version}.json"
        with open(rule_file, 'w', encoding='utf-8') as f:
            json.dump(rule.model_dump(), f, ensure_ascii=False, indent=2, default=str)

    def get_rule(self, version: Optional[str] = None) -> PortInspectionRule:
        if version is None:
            return self.get_latest_rule()
        rule_file = self.rules_dir / f"rules_{version}.json"
        if not rule_file.exists():
            raise ValueError(f"规则版本 {version} 不存在")
        with open(rule_file, 'r', encoding='utf-8') as f:
            data = json.load(f)
        return PortInspectionRule(**data)

    def get_latest_rule(self) -> PortInspectionRule:
        rule_files = sorted(self.rules_dir.glob("rules_*.json"))
        if not rule_files:
            raise ValueError("没有找到任何规则文件")
        latest_file = rule_files[-1]
        with open(latest_file, 'r', encoding='utf-8') as f:
            data = json.load(f)
        return PortInspectionRule(**data)

    def list_rule_versions(self) -> List[RuleVersion]:
        versions = []
        for rule_file in sorted(self.rules_dir.glob("rules_*.json")):
            with open(rule_file, 'r', encoding='utf-8') as f:
                data = json.load(f)
            versions.append(RuleVersion(
                version=data['version'],
                effective_date=datetime.fromtimestamp(rule_file.stat().st_mtime),
                description=data['description'],
                is_active=True
            ))
        return versions

    def get_port_risk_level(self, port: int, rule_version: Optional[str] = None) -> RiskLevel:
        rule = self.get_rule(rule_version)
        for port_range, risk_level in rule.port_ranges.items():
            start, end = map(int, port_range.split('-'))
            if start <= port <= end:
                if port in rule.reserved_ports:
                    return RiskLevel.CRITICAL
                return risk_level
        return RiskLevel.LOW

    def is_high_connections(self, connection_count: int, rule_version: Optional[str] = None) -> bool:
        rule = self.get_rule(rule_version)
        return connection_count >= rule.threshold_connections

File: zy70458/src/main/java/com/portinspector/RuleManager.java
package com.portinspector;

import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.datatype.jsr310.JavaTimeModule;
import com.portinspector.model.PortInspectionRule;
import com.portinspector.model.RiskLevel;

import java.io.File;
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.*;
import java.util.stream.Collectors;

public class RuleManager {
    private final Path rulesDir;
    private final ObjectMapper objectMapper;

    public RuleManager(String dataDir) {
        this.rulesDir = Paths.get(dataDir, "rules");
        this.objectMapper = new ObjectMapper();
        this.objectMapper.registerModule(new JavaTimeModule());
        initDefaultRules();
    }

    private void initDefaultRules() {
        try {
            Files.createDirectories(rulesDir);

            Path rule1 = rulesDir.resolve("rules_v1.0.0.json");
            if (!Files.exists(rule1)) {
                PortInspectionRule v1 = new PortInspectionRule();
                v1.setVersion("v1.0.0");
                Map<String, RiskLevel> ranges1 = new LinkedHashMap<>();
                ranges1.put("0-1023", RiskLevel.CRITICAL);
                ranges1.put("1024-49151", RiskLevel.MEDIUM);
                ranges1.put("49152-65535", RiskLevel.LOW);
                v1.setPortRanges(ranges1);
                v1.setReservedPorts(Arrays.asList(22, 80, 443, 3306, 5432, 6379, 27017));
                v1.setThresholdConnections(100);
                v1.setDescription("初始版本规则：系统端口CRITICAL，注册端口MEDIUM，动态端口LOW");
                objectMapper.writeValue(rule1.toFile(), v1);
            }

            Path rule2 = rulesDir.resolve("rules_v2.0.0.json");
            if (!Files.exists(rule2)) {
                PortInspectionRule v2 = new PortInspectionRule();
                v2.setVersion("v2.0.0");
                Map<String, RiskLevel> ranges2 = new LinkedHashMap<>();
                ranges2.put("0-1023", RiskLevel.HIGH);
                ranges2.put("1024-49151", RiskLevel.LOW);
                ranges2.put("49152-65535", RiskLevel.SAFE);
                v2.setPortRanges(ranges2);
                v2.setReservedPorts(Arrays.asList(22, 80, 443, 3306, 5432, 6379, 27017, 8080, 8443));
                v2.setThresholdConnections(50);
                v2.setDescription("v2.0版本：降低系统端口等级为HIGH，注册端口为LOW，新增常用端口");
                objectMapper.writeValue(rule2.toFile(), v2);
            }
        } catch (IOException e) {
            throw new RuntimeException("初始化规则文件失败", e);
        }
    }

    public PortInspectionRule getRule(String version) {
        try {
            Path ruleFile = rulesDir.resolve("rules_" + version + ".json");
            return objectMapper.readValue(ruleFile.toFile(), PortInspectionRule.class);
        } catch (IOException e) {
            throw new RuntimeException("规则版本不存在: " + version, e);
        }
    }

    public PortInspectionRule getLatestRule() {
        try {
            List<Path> ruleFiles = Files.list(rulesDir)
                    .filter(p -> p.getFileName().toString().startsWith("rules_"))
                    .sorted()
                    .collect(Collectors.toList());
            if (ruleFiles.isEmpty()) {
                throw new RuntimeException("没有找到任何规则文件");
            }
            return objectMapper.readValue(ruleFiles.get(ruleFiles.size() - 1).toFile(), PortInspectionRule.class);
        } catch (IOException e) {
            throw new RuntimeException("获取最新规则失败", e);
        }
    }

    public List<String> listRuleVersions() {
        try {
            return Files.list(rulesDir)
                    .filter(p -> p.getFileName().toString().startsWith("rules_"))
                    .map(p -> p.getFileName().toString().replace("rules_", "").replace(".json", ""))
                    .sorted()
                    .collect(Collectors.toList());
        } catch (IOException e) {
            return Collections.emptyList();
        }
    }

    public RiskLevel getPortRiskLevel(int port, String ruleVersion) {
        PortInspectionRule rule = getRule(ruleVersion);
        if (rule.getReservedPorts().contains(port)) {
            return RiskLevel.CRITICAL;
        }
        for (Map.Entry<String, RiskLevel> entry : rule.getPortRanges().entrySet()) {
            String[] parts = entry.getKey().split("-");
            int start = Integer.parseInt(parts[0]);
            int end = Integer.parseInt(parts[1]);
            if (port >= start && port <= end) {
                return entry.getValue();
            }
        }
        return RiskLevel.LOW;
    }

    public boolean isHighConnections(int connectionCount, String ruleVersion) {
        return connectionCount >= getRule(ruleVersion).getThresholdConnections();
    }
}
